feat: add Vue composable package and introduced a vue-supabase-todo project

Chriztiaan opened the pull request:
# Description
Added the vue package which make [composables](https://vuejs.org/guide/reusability/composables) available for accessing the powerSync instance, static queries, and watched queries.
Also created a `vue-supabase-todolist` demo project.

## Screenshots
| Vue | React |
|--------|--------|
| ![vue_lists](https://github.com/powersync-ja/powersync-js/assets/13052031/1f50f872-13dc-4bad-b20b-ea66859f676f) | ![react_lists](https://github.com/powersync-ja/powersync-js/assets/13052031/4b8f842b-cbca-46ac-91fa-3a2ad720d113) |
| ![vue_todos](https://github.com/powersync-ja/powersync-js/assets/13052031/b8e60803-e082-43c8-8ebf-4917029b0357) | ![react_todos](https://github.com/powersync-ja/powersync-js/assets/13052031/f77ccfc5-4aba-47e5-b7fe-11169f63b294) |
| ![vue_console](https://github.com/powersync-ja/powersync-js/assets/13052031/c06a3223-4de4-4a6e-b90c-cc05f5265b51) | ![react_console](https://github.com/powersync-ja/powersync-js/assets/13052031/42a613f6-0125-4d3c-b666-57fa6b4a1236) |
| ![vue_login](https://github.com/powersync-ja/powersync-js/assets/13052031/6d34514a-b51b-4981-bca7-ca8324c47e89) | ![react_login](https://github.com/powersync-ja/powersync-js/assets/13052031/143136bb-2f03-4003-bcaa-229d7a8f1574) |
| ![error](https://github.com/powersync-ja/powersync-js/assets/13052031/8c6bb509-ad03-4c08-b994-d5ebf79d5e94)
 | |

stevensJourney commented on demos/vue-supabase-todolist/components.d.ts:
```diff
@@ -0,0 +1,27 @@
+/* eslint-disable */
```
Should this file be comited?

stevensJourney commented on demos/vue-supabase-todolist/src/App.vue:
```diff
@@ -0,0 +1,29 @@
+
+powerSync.init();
+supabase.registerListener({
+  initialized: () => {},
```
The `initialized` listener should not be required. Should be fine to omit empty handler

stevensJourney commented on demos/vue-supabase-todolist/src/plugins/powerSync.ts:
```diff
@@ -0,0 +1,10 @@
+import { createPowerSync } from '@journeyapps/powersync-vue';
+
+export const powerSync = new WASQLitePowerSyncDatabaseOpenFactory({
+  dbFilename: 'example.db',
```
Can we give this a more unique name which matches the demo? This can make debugging easier when running multiple demo apps.

stevensJourney commented on packages/powersync-vue/src/composables/usePowerSyncQuery.ts:
```diff
@@ -0,0 +1,56 @@
+      error.value = undefined;
+      loading.value = true;
+
+      await powerSync.value.readLock(async (tx) => {
+        const result = await tx.getAll(toValue(sqlStatement), toValue(parameters));
+        data.value = result;
+      });
```
```suggestion
      data.value = await powerSync.value.getAll(toValue(sqlStatement), toValue(parameters));
```

The `getAll` should use a readLock by default.

stevensJourney commented on packages/powersync-vue/src/composables/usePowerSyncQuery.ts:
```diff
@@ -0,0 +1,56 @@
+import { MaybeRef, Ref, ref, toValue, watch } from 'vue';
+import { usePowerSync } from './powerSync';
+
+export type QueryOptions = { watchParameters: boolean; immediate: boolean };
```
Please add some comments which describe what these options do. E.g `immediate` isn't very descriptive of what is immediate.

stevensJourney commented on packages/powersync-vue/src/composables/usePowerSyncWatchedQuery.ts:
```diff
@@ -0,0 +1,69 @@
+      return;
+    }
+
+    powerSync.value.watch(
```
This might depend on the concept of `loading` versus `fetching`. But if you used `powersync.onchange` here instead as a trigger then manually executed the query then you could set either `loading/fetching` to true for each time the query is actually executed/refreshed.

Chriztiaan replied:
That sounds granular, which I kinda like. But the suggestion I am working from was a loading state which locks to false when the first result/error comes.

Chriztiaan replied:
Is it worth having exposing them as two different variables so that we track both, or do you think we should condense it into the same variable?

Chriztiaan replied:
We now both `loading` for hard loading and `fetching` which indicated whether the query is being evaluated.

stevensJourney replied:
Awesome!

benitav commented on demos/vue-supabase-todolist/README.md:
```diff
@@ -0,0 +1,60 @@
+
+Demo app demonstrating use of the [PowerSync Vue](https://www.npmjs.com/package/@journeyapps/powersync-vue) together with Supabase.
+
+A step-by-step guide on Supabase<>PowerSync integration is available [here](https://docs.powersync.com/integration-guides/supabase).
```
```suggestion
A step-by-step guide through the Supabase<>PowerSync integration is available [here](https://docs.powersync.com/integration-guides/supabase).
```

benitav commented on packages/powersync-vue/README.md:
```diff
@@ -0,0 +1,157 @@
+
+The `createPowerSyncPlugin` function is designed for setting up a PowerSync client that is available across your entire Vue application. It's the recommended approach for package setup. However, there may be situations where an app-wide setup isn't suitable, or you need a different PowerSync client for specific parts of your application.
+
+In these cases, you can use the `providePowerSync` function within a parent component to override the PowerSync instance for all its descendant components. This allows for more granular control over the which PowerSync client is used in different sections of your application. `providePowerSync` can be utilized regardless of whether `createPowerSyncPlugin` has been used globally. If there are mult ...
```
```suggestion
In these cases, you can use the `providePowerSync` function within a parent component to override the PowerSync instance for all its descendant components. This allows for more granular control over which PowerSync client is used in different sections of your application. `providePowerSync` can be utilized regardless of whether `createPowerSyncPlugin` has been used globally. If there are multiple uses of `providePowerSync` within a component hierarchy, the closest `providePowerSync` call to the composable invocation will determine the PowerSync instance used.
```

benitav commented on packages/powersync-vue/README.md:
```diff
@@ -0,0 +1,157 @@
+
+## Setup
+
+To set up app-wide accessibility of PowerSync composables, create a PowerSync Vue plugin that configures a PowerSync instance and integrates it with the Vue instance. This approach provides the PowerSync client app wide so that any composable used has access to the instance.
```
The second sentence here is difficult to read. Is it necessary to keep it, trying to read it it sounds like it's saying more or less what the previous sentence says?

benitav commented on packages/powersync-vue/README.md:
```diff
@@ -0,0 +1,157 @@
+
+### Queries
+
+The `usePowerSyncQuery` composable provides a static view of a given query, but can you use refs as parameters instead to automatically refresh the query when they change. The composable exposes reactive variables for the results, the loading state and error state, as well as a refresh callback that can be invoked to rerun the query manually.
```
```suggestion
The `usePowerSyncQuery` composable provides a static view of a given query. You can use refs as parameters instead to automatically refresh the query when they change. The composable exposes reactive variables for the results, the loading state and error state, as well as a refresh callback that can be invoked to rerun the query manually.
```

benitav commented on packages/powersync-vue/README.md:
```diff
@@ -0,0 +1,157 @@
+
+### Watched Queries
+
+The `usePowerSyncWatchedQuery` composable provides a dynamic view of a given query, the data will automatically update when a dependant table is updated.
```
```suggestion
The `usePowerSyncWatchedQuery` composable provides a dynamic view of a given query. The data will automatically update when a dependent table is updated.
```

benitav commented on packages/powersync-vue/README.md:
```diff
@@ -0,0 +1,157 @@
+The `usePowersync`, `usePowerSyncQuery`, and `usePowerSyncWatchedQuery` composables are meant to be invoked in the top-level setup block. Vue expects certain Composition API functions, like `inject` which this package depends on, to be resolved in the setup context and not inside nested or asynchronous functions. For use cases where you need to do this, you should access the PowerSync `AbstractPo ...
+
+Incorrect Usage Example:
+Using powerSync composables in a nested function of a component.
```
```suggestion
Using PowerSync composables in a nested function of a component.
```

Chriztiaan commented:
Updated the watched hook to have a `fetching` state alongside the `loading` state:
![image](https://github.com/powersync-ja/powersync-js/assets/13052031/263cedab-fd6c-4ad2-86fc-6dfb4e6aaae8)

Introduced a connectivity composable, which uses the new `await db.waitForFirstSync()`- which I am note sure how we want to document.
![image](https://github.com/powersync-ja/powersync-js/assets/13052031/053ed0a8-8203-45e7-a827-280b07d7c5d1)

Chriztiaan commented on packages/powersync-sdk-common/src/client/AbstractPowerSyncDatabase.ts:
```diff
@@ -852,10 +910,7 @@ export abstract class AbstractPowerSyncDatabase extends BaseObserver<PowerSyncDB
-   * @ignore
-   */
-  private async executeReadOnly(sql: string, params?: any[]) {
+  async executeReadOnly(sql: string, params?: any[]) {
```
@stevensJourney is this being public acceptable? I needed it in the new watched hook that depends on `onChange` instead of `watch`.

benitav commented on demos/vue-supabase-todolist/README.md:
```diff
@@ -0,0 +1,64 @@
+
+## Note: Alpha Release
+
+This package is currently in a alpha release.
```
```suggestion
This package is currently in an alpha release.
```

stevensJourney commented on packages/powersync-vue/src/composables/usePowerSyncStatus.ts:
```diff
@@ -0,0 +1,38 @@
+export const usePowerSyncStatus = () => {
+  const status = ref(
+    new SyncStatus({
+      connected: true,
```
Maybe this should be false initially?

rkistner commented on demos/vue-supabase-todolist/src/components/LoadingMessage.vue:
```diff
@@ -0,0 +1,22 @@
+  fetching: Boolean
+});
+
+const loadingMessage = computed(() => (props.loading ? 'Busy with initial sync' : 'Re-executing query'));
```
"Busy with initial sync" is not accurate here - this is displayed for any loading state.

rkistner replied:
The loader is often displayed very briefly when fetching or re-fetching data. Is there a nice way to only display it when fetching takes more than 300ms for example?

This is more of an issue for re-fetching than the initial load, since it happens much more often.

In some cases the query is fast enough that it's not visible, but it's easily seen if we add e.g. a 30ms delay to queries.

rkistner commented on packages/powersync-sdk-common/src/client/AbstractPowerSyncDatabase.ts:
```diff
@@ -132,25 +132,40 @@ export abstract class AbstractPowerSyncDatabase extends BaseObserver<PowerSyncDB
     super();
     this.bucketStorageAdapter = this.generateBucketStorageAdapter();
     this.closed = false;
     this.currentStatus = undefined;
```
@stevensJourney Could we populate the initial status here? Currently it seems to only be populated when first connected, which means that the vue `usePowerSyncStatus` needs to infer it's own initial status until then.

rkistner commented on packages/powersync-sdk-common/src/client/AbstractPowerSyncDatabase.ts:
```diff
@@ -132,25 +132,40 @@ export abstract class AbstractPowerSyncDatabase extends BaseObserver<PowerSyncDB
+  /**
+   * Indicates whether there has been at least one full sync.
+   */
+  hasSynced: boolean;
```
Can we move this to the SyncStatus, and use the same mechanism to notify listeners of changes? That could simplify `usePowerSyncStatus` a bit.

I'd also suggest having an additional `undefined` state to mean "unknown" (i.e. still loading the state from the database when loading the page).

The `waitForFirstSync()` would still be useful regardless of the above change.

rkistner commented on packages/powersync-vue/src/composables/usePowerSyncWatchedQuery.ts:
```diff
@@ -0,0 +1,97 @@
+
+    try {
+      // Fetch initial data
+      onResult(await powerSync.value.getAll<T>(sql, parameters));
+
+      const resolvedTables = await powerSync.value.resolveTables(sql, parameters, options);
+
+      powerSync.value.onChangeWithCallback(
```
There is a race condition possibility here: There could be changes after calling `getAll()`, but before calling `onChangeWithCallback`.

This is also an issue with `AbstractPowerSyncDatabase.watch()`, so it's okay to leave as is for now and fix it all in a separate PR.

Chriztiaan replied:
`AbstractPowerSyncDatabase.watch` is fixed here https://github.com/powersync-ja/powersync-js/pull/120.
Will apply that fix in this file too.

rkistner commented on demos/vue-supabase-todolist/src/components/widgets/TodoListsWidget.vue:
```diff
@@ -0,0 +1,65 @@
+<template>
+  <div class="py-10 pa-2">
+    <LoadingMessage v-if="loading || fetching" :loading :fetching />
```
Could we also use hasSynced here? Would be great do demonstrating waiting for initial sync.

stevensJourney commented on packages/powersync-sdk-common/src/client/AbstractPowerSyncDatabase.ts:
```diff
@@ -209,10 +237,42 @@ export abstract class AbstractPowerSyncDatabase extends BaseObserver<PowerSyncDB
+          const hasSynced = !!result.rows?.length;
+          if (hasSynced) {
+            abortController.abort();
+            this.currentStatus = new SyncStatus({ ...this.currentStatus.toJSON(), hasSynced });
```
This doesn't set `hasSynced` to `false`

stevensJourney commented on packages/powersync-vue/src/composables/usePowerSyncStatus.ts:
```diff
@@ -0,0 +1,29 @@
+import { usePowerSync } from './powerSync';
+
+export const usePowerSyncStatus = () => {
+  const status = ref(new SyncStatus({}));
```
You should be able to use the status of the PowerSync client here now

Chriztiaan replied:
This is just a default until the powerSync instance is resolved on line 11.

stevensJourney reviewed (dismissed):
Looks good from my side. Happy to release this.

DominicGBauer commented on packages/powersync-sdk-common/src/client/AbstractPowerSyncDatabase.ts:
```diff
@@ -193,6 +197,30 @@ export abstract class AbstractPowerSyncDatabase extends BaseObserver<PowerSyncDB
 
+  /**
+   * @returns A promise which will resolve once the first full sync has completed.
+   */
```
Can you add an example implementation of this

Chriztiaan replied:
This would be an example usage:
```
export default function SyncStatus() {
  const powerSync = usePowerSync();

  const [hasSynced, setHasSynced] = useState(false);
  useEffect(() => {
    async function handleSyncState() {
      await powerSync.waitForFirstSync();
      setHasSynced(true);
    }
    handleSyncState();
  }, []);

  return <div>{hasSynced ? 'Initial sync completed!' : 'Busy with initial sync...'}</div>;
}
```

But since it's for a specific framework (react) - I don't think I could add it to the common SDK's docs.

benitav replied:
We will be adding examples to our docs, e.g. under https://docs.powersync.com/client-sdk-references/react-native-and-expo/usage-examples

DominicGBauer replied:
@benitav it would be great to have example in the JSDoc section as this is what will be shown to users who hover over the function (which can be super useful) as well as add an example when JS Docs are generated e.g. 
<img width="693" alt="image" src="https://github.com/powersync-ja/powersync-js/assets/46312751/b055814e-ab4e-4a11-8615-3204f576870d">

<img width="861" alt="image" src="https://github.com/powersync-ja/powersync-js/assets/46312751/a6dcc53e-ed57-482f-8591-30384f099b0d">

benitav commented:
Yes, I agree. But per Christiaan's comment it sounds like this is not
possible without framework-specific code. I'll defer to him/Steven on that,
I just mentioned the docs to share that we are working on an example
implementation.

On Tue, Apr 16, 2024 at 2:12 PM Dominic Gunther Bauer <
***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In packages/powersync-sdk-common/src/client/AbstractPowerSyncDatabase.ts
> <https://github.com/powersync-ja/powersync-js/pull/111#discussion_r1567250788>
> :
>
> > @@ -193,6 +197,30 @@ export abstract class AbstractPowerSyncDatabase extends BaseObserver<PowerSyncDB
>      await this._isReadyPromise;
>    }
>
> +  /**
> +   * @returns A promise which will resolve once the first full sync has completed.
> +   */
>
> @benitav <https://github.com/benitav> it would be great to have example
> in the JSDoc section as this is what will be shown to users who hover over
> the function (which can be super useful) as well as add an example when JS
> Docs are generated e.g.
> image.png (view on web)
> <https://github.com/powersync-ja/powersync-js/assets/46312751/b055814e-ab4e-4a11-8615-3204f576870d>
> image.png (view on web)
> <https://github.com/powersync-ja/powersync-js/assets/46312751/a6dcc53e-ed57-482f-8591-30384f099b0d>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/powersync-ja/powersync-js/pull/111#discussion_r1567250788>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABYH5IEIF2RTPLDAU7MPZZDY5UIUJAVCNFSM6AAAAABFVQ2A3SVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDAMBTGQ2TENRVHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
